File: java/pe/src/main/java/ktn/pe/Pe1.java
package ktn.pe;

import ktn.pe.util.PeUtils;

public class Pe1 implements Pe {
    private int limit = 1000;

    /**
     * double sum of 0, 1, ..., limit
     * 
     * @param limit
     * @return
     */
    private int sumToX2(int limit) {
        return limit * (limit + 1);
    }

    /**
     * 
     * @param limit
     * @return
     */
    public int pe1(int limit) {
        int sum = 0;

        if (limit > 2) {
            sum = (3 * sumToX2(limit / 3) + 5 * sumToX2(limit / 5) - 15 * sumToX2(limit / 15)) >> 1;
        }

        return sum;
    }

    /**
     * 
     * @param limit
     * @return
     */
    public int pe1loop(int limit) {
        int sum = 0;

        for (int i = 3; i <= limit; ++i) {
            if (i % 3 == 0 || i % 5 == 0) {
                sum += i;
            }
        }

        return sum;
    }

    @Override
    public void setArgs(String[] args) {
        if (args.length > 0) {
            limit = Integer.parseInt(args[0]);
        }
    }

    @Override
    public void solve() {
        System.out.println(PeUtils.format(this, pe1(limit)));
    }

    @Override
    public void run() {
        solve();
    }

    @Override
    public int getProblemNumber() {
        return 1;
    }
}

File: kt/pe/src/main/kotlin/ktn/pe/Pe1.kt
package ktn.pe

class Pe1 : Pe {
    private var limit = 1000

    override val problemNumber: Int
        get() = 1

    /**
     * double sum of 0, 1, ..., limit
     *
     * @param limit
     * @return
     */
    private fun sumToX2(limit: Int): Int {
        return limit * (limit + 1)
    }

    /**
     *
     * @param limit
     * @return
     */
    fun pe1(limit: Int): Int {
        var sum = 0

        if (limit > 2) {
            sum = (3 * sumToX2(limit / 3)
                + 5 * sumToX2(limit / 5)
                - 15 * sumToX2(limit / 15)) shr 1
        }

        return sum
    }

    /**
     *
     * @param limit
     * @return
     */
    fun pe1loop(limit: Int): Int {
        return (3..limit)
                .filter { it % 3 == 0 || it % 5 == 0 }
                .sum()
    }

    override fun setArgs(args: Array<String>) {
        if (args.isNotEmpty()) {
            limit = args[0].toInt()
        }
    }

    override fun solve() {
        println(PeUtils.format(this, pe1(limit)))
    }

    override fun run() {
        solve()
    }
}
